### ISSUE TITLE
read.adp.ad2cp() ought to be vectorized

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1980]
Vectorizing should speed up processing, and simplify the code.  From #1967, I think vectorization is always feasible.

I'll do this in the `ad2cp` branch, ticking off items as I do them, but only pushing to GH when I feel like I need a steppingstone.  Also, to the annoyance of others, I plan to take most of my notes by editing the present comment, because scrolling is a bother.

Note to self: bookmark `J` is where the old unvectorized code lives, and `A` is at 'burst' (which I use as a guide to coding other things) and `B` is at whatever I'm working on presently. 

Note to others: I will start with ID types that are contained within files that are in my possession (marked :paperclip: below).  When -- or if -- I move on to other ID types is an open question, for guessing with any way to test even basic assumptions is not a good use of time.

* [x] :paperclip: `burst                    (ID=0x15=21)` - new configuration system
* [x] :paperclip:  `average               (ID=0x16=22)` - new configuration system
* [x] :paperclip: `bottomTrack       (ID=0x17=23)` - new configuration system;  see #1982
* [x] :paperclip:  #1984
* [x] :paperclip:  `burstAltimeterRaw (ID=0x1a=26)` - old configuration system
* [x] `DVLBottomTrack    (ID=0x1b=27)`
* [x] :paperclip: #1983. (Done, but leaving 1983 open for discussion with @richardsc)
* [x] `DVLWaterTrack     (ID=0x1d=29)`
* [x] `altimeter         (ID=0x1e=30)`
* [x] `averageAltimeter  (ID=0x1f=31)`
* [x] :paperclip: `text              (ID=0xa0=160)` - old configuration system

**Comment by dankelley** [comment]
Here are the files I'm using for testing.
```R
> read.oce("tests/testthat/local_data/ad2cp/ad2cp_01.ad2cp",which="?")
    0x16=average 0x1c=echosounder           0x23=?        0xa0=text 
               7                2                2                1 
```
```R
> read.oce("tests/testthat/local_data/ad2cp/S102791A002_Barrow_v2.ad2cp",which="?")
            0x15=burst           0x16=average       0x17=bottomTrack 
                   648                    360                    360 
0x1a=burstAltimeterRaw              0xa0=text 
                   108                      1 
```

```R
> read.oce("/Users/kelley/Dropbox/oce_secret_data//ad2cp_secret_1.ad2cp",which="?")
  0x15=burst 0x16=average    0xa0=text 
      235016        29377            1 
```

```R
> read.oce("/Users/kelley/Dropbox/oce_secret_data//ad2cp_secret_2.ad2cp",which="?")
0x15=burst  0xa0=text 
  10419583          1 
```

```R
> read.oce("/Users/kelley/Dropbox/oce_secret_data//ad2cp_secret_3.ad2cp",which="?")
           0x15=burst 0x18=interleavedBurst             0xa0=text 
                 5551                  5550                     1 
```

**Comment by dankelley** [comment]
NOTE: the `configuration` set differs from one ID type to another.

<details>

**Burst/Average** (Nortek 2017 Table 6.1.2 page 48)

<img width="1036" alt="Screenshot 2022-07-17 at 9 27 32 AM" src="https://user-images.githubusercontent.com/99469/179398435-d1ec7239-f4d3-47d0-bd2b-b3e7203175b3.png">


**Bottom Track** (Nortek 2017 Table 6.1.3 page 60)

<img width="1036" alt="Screenshot 2022-07-17 at 9 26 46 AM" src="https://user-images.githubusercontent.com/99469/179398461-d6957229-3781-400b-add2-32bd39706003.png">

**Comment by dankelley** [comment]
I'm migrating to what I call 'tailored vectorization'.  This is necessary because my old method of reading a whole bunch of things at once, assume this will work with all data types, is simply false. 

See Nortek (2022 page 80), reproduced below ... the consequence is that all data coming after "ambiguity velocity" is at one of two positions, depending on the type.  So, things above that in the table work with the old code, but not things after that.  And data start at byte 77 for some types, and 79 for other types ... in contradiction to my old assumption that things started at 77.  This does not matter for existing users, because I had not coded bottom-track yet... that's next.



<img width="612" alt="Screenshot 2022-07-19 at 9 27 16 AM" src="https://user-images.githubusercontent.com/99469/179750065-dba51c59-191f-441e-a121-78ae96386b86.png">

**Comment by dankelley** [comment]
I've vectorized all of the IDs in commit fe34c984687a854a66485375c0f5ed7a34b654db  of the "ad2cp" branch.  I'm not saying that I know that they work, though, for (a) I only have test files for a few IDs, and (b) the docs say next to nothing about some types.  But, at least the code is not now processing things one chunk at a time, which would be pretty inefficient for say a 1G file holding 1K chunks (a million separate calls to `readBin()` is just the beginning; for each chunk, we need to read multiple things, so my guess would be maybe 5M calls for a 1G file).

Since the issue -- as stated by it's title -- has been addressed, I am closing this.  If some things proved to be incorrect, there should be finer-grained issues on them, so solutions will be clearer to me and to users.
